File: lib/iriq/evidence.rb
module Iriq
  # Evidence is the structured substrate for explanation. Each Record
  # captures one fact about the system's reasoning: "this segment
  # classified as :integer because the Integer recognizer fired with
  # specificity TYPED", "the IPv4 type collapses to {ip} by policy",
  # "Position P is mostly variable because of corpus stats".
  #
  # Trace and Explanation are views over a list of Evidence records;
  # the structured form is what programmatic consumers (test assertions,
  # PR-diff annotators, downstream tooling) should build on. Human note
  # strings emitted by Trace are derived from Evidence payloads, so
  # adding a new note kind starts with adding a new Evidence shape.
  #
  # Two axes:
  #
  #   subject_kind ∈ {:segment, :position, :cluster}
  #     What this Evidence is about. Today most Evidence is :segment
  #     (per-segment classification facts). :position and :cluster
  #     Evidence become load-bearing once corpus-informed Trace lands
  #     in a follow-up step.
  #
  #   source ∈ {:lexical, :recognizer, :corpus, :neighbor, :policy}
  #     What kind of fact is being asserted.
  #       :lexical    — pure shape match (e.g. "matches DATE_RE")
  #       :recognizer — a named Recognizer fired with confidence/specificity
  #       :corpus     — aggregated counts/distributions support this
  #       :neighbor   — adjacent context informed this (prior literal,
  #                     param name hint)
  #       :policy     — a normalization policy applied (ip umbrella
  #                     collapse, canonical date, currency upcase)
  module Evidence
    SUBJECT_KINDS = %i[segment position cluster].freeze
    SOURCES       = %i[lexical recognizer corpus neighbor policy].freeze

    # A single Evidence fact.
    #
    #   subject_kind — :segment | :position | :cluster
    #   subject      — kind-specific identity:
    #                    :segment  → { index:, value: }
    #                    :position → Iriq::Position
    #                    :cluster  → cluster key (string)
    #   source       — :lexical | :recognizer | :corpus | :neighbor | :policy
    #   payload      — source-and-kind-specific structured data
    #   weight       — optional float in [0,1] — contribution to the
    #                  ultimate decision. Set when scoring is meaningful;
    #                  nil otherwise.
    #   notes        — optional human-readable strings. Trace renders
    #                  these directly; programmatic consumers can ignore.
    class Record
      attr_reader :subject_kind, :subject, :source, :payload, :weight, :notes

      def initialize(subject_kind:, subject:, source:, payload:, weight: nil, notes: [])
        unless SUBJECT_KINDS.include?(subject_kind)
          raise ArgumentError, "subject_kind must be one of #{SUBJECT_KINDS.inspect}"
        end
        unless SOURCES.include?(source)
          raise ArgumentError, "source must be one of #{SOURCES.inspect}"
        end

        @subject_kind = subject_kind
        @subject      = subject
        @source       = source
        @payload      = payload || {}
        @weight       = weight
        @notes        = notes || []
      end

      def to_h
        {
          subject_kind: @subject_kind,
          subject:      subject_serialized,
          source:       @source,
          payload:      @payload,
          weight:       @weight,
          notes:        @notes,
        }.compact
      end

      private

      def subject_serialized
        return @subject.to_h if @subject.respond_to?(:to_h) && !@subject.is_a?(Hash)
        @subject
      end
    end

    module_function

    # Factories so call sites don't have to repeat subject_kind:.
    def segment(index:, value:, source:, payload:, weight: nil, notes: [])
      Record.new(
        subject_kind: :segment,
        subject:      { index: index, value: value },
        source:       source, payload: payload, weight: weight, notes: notes,
      )
    end

    def position(position:, source:, payload:, weight: nil, notes: [])
      Record.new(
        subject_kind: :position,
        subject:      position,
        source:       source, payload: payload, weight: weight, notes: notes,
      )
    end

    def cluster(key:, source:, payload:, weight: nil, notes: [])
      Record.new(
        subject_kind: :cluster,
        subject:      key,
        source:       source, payload: payload, weight: weight, notes: notes,
      )
    end
  end
end

File: go/evidence.go
package iriq

import "fmt"

// EvidenceSubjectKind distinguishes what an Evidence record is about.
type EvidenceSubjectKind string

const (
	EvidenceSubjectSegment  EvidenceSubjectKind = "segment"
	EvidenceSubjectPosition EvidenceSubjectKind = "position"
	EvidenceSubjectCluster  EvidenceSubjectKind = "cluster"
)

// EvidenceSource distinguishes what kind of fact an Evidence record asserts.
//
//   EvidenceLexical    — pure shape match (e.g. matches a regex)
//   EvidenceRecognizer — a named Recognizer fired with confidence/specificity
//   EvidenceCorpus     — aggregated counts/distributions support this
//   EvidenceNeighbor   — adjacent context informed this (prior literal,
//                        param name hint)
//   EvidencePolicy     — a normalization policy applied (ip umbrella
//                        collapse, canonical date, currency upcase)
type EvidenceSource string

const (
	EvidenceLexical    EvidenceSource = "lexical"
	EvidenceRecognizer EvidenceSource = "recognizer"
	EvidenceCorpus     EvidenceSource = "corpus"
	EvidenceNeighbor   EvidenceSource = "neighbor"
	EvidencePolicy     EvidenceSource = "policy"
)

// Evidence is the structured substrate for explanation.
//
// Each record captures one fact about the system's reasoning: "this segment
// classified as :integer because the Integer recognizer fired", "the IPv4
// type collapses to {ip} by policy", "Position P is mostly variable because
// of corpus stats".
//
// Trace and Explanation are views over Evidence; the structured form is
// what programmatic consumers should build on. Note strings emitted by
// Trace derive from Evidence payloads.
//
//   SubjectKind = segment   Subject is EvidenceSegmentSubject (index + value)
//   SubjectKind = position  Subject is Position
//   SubjectKind = cluster   Subject is the cluster key (string)
//
// Weight is in [0, 1] when meaningful; zero otherwise. Notes are
// human-readable strings — Trace renders them verbatim.
type Evidence struct {
	SubjectKind EvidenceSubjectKind
	Subject     any
	Source      EvidenceSource
	Payload     map[string]any
	Weight      float64
	Notes       []string
}

// EvidenceSegmentSubject identifies a single segment within an Identifier.
// Index is the integer position for path segments and the param name
// (as any) for query params — keeps wire compatibility with Ruby.
type EvidenceSegmentSubject struct {
	Index any    // int for path segments, string for query params
	Value string
}

// SegmentEvidence is a constructor for segment-kind Evidence records.
func SegmentEvidence(index any, value string, source EvidenceSource, payload map[string]any) Evidence {
	return Evidence{
		SubjectKind: EvidenceSubjectSegment,
		Subject:     EvidenceSegmentSubject{Index: index, Value: value},
		Source:      source,
		Payload:     payload,
	}
}

// PositionEvidence is a constructor for position-kind Evidence records.
func PositionEvidence(pos Position, source EvidenceSource, payload map[string]any) Evidence {
	return Evidence{
		SubjectKind: EvidenceSubjectPosition,
		Subject:     pos,
		Source:      source,
		Payload:     payload,
	}
}

// ClusterEvidence is a constructor for cluster-kind Evidence records.
func ClusterEvidence(key string, source EvidenceSource, payload map[string]any) Evidence {
	return Evidence{
		SubjectKind: EvidenceSubjectCluster,
		Subject:     key,
		Source:      source,
		Payload:     payload,
	}
}

// WithNotes returns a copy of the Evidence with the given notes appended.
func (e Evidence) WithNotes(notes ...string) Evidence {
	e.Notes = append(append([]string{}, e.Notes...), notes...)
	return e
}

// WithWeight returns a copy of the Evidence with weight set.
func (e Evidence) WithWeight(w float64) Evidence {
	e.Weight = w
	return e
}

func (e Evidence) String() string {
	return fmt.Sprintf("Evidence(%s/%s, %v)", e.SubjectKind, e.Source, e.Payload)
}
